### ISSUE TITLE
[EPCO-6] Search actions and states must be defined

### ISSUE THREAD

**Issue by hatfield-c** [hatfield-c/epco#6]
- A search algorithm will need to be settled on and/or defined
    - A* is simple, but requires a lot of memory (which might be acceptable)
    - IDA* is supposed to work well with changing environments
    - RBFS (Recursive Best First Search?) can potentially be extended to include more than one alternative state?
- Special care must be paid to how states are defined during the search procedure
- Heuristic function will need to use N in some way
    - The working theory is that the gradient of N at certain positions will point towards the closest wall.
    - The previous would imply that the Hessian of N could also guide search, as it would tell us how the direction to the nearest wall is changing, from which we can make inferences about the local structure.

**Comment by hatfield-c** [comment]
When a state is entered, it must be valid. When an action is taken, the resultant state must be valid. When the timestep transitions, the following state must be valid.

Thus so long as we ensure that we begin in a valid state, we only need to check the action-result condition and the timestep result condition each update cycle. The problem with this is that Unity is not designed for backwards stepping of simulation frames. 

Some features will be pre-encoded and can be looked at in advanced. Other applications (i.e. wind in aeronautics) will be non-deterministic. For now, such shortcomings will have to be accepted. They might be mitigated at least by restricting the domain in some way (i.e. wind is at or near 13km/h +/- 3km/h )

**Comment by hatfield-c** [comment]
Non-determinism is not addressed by this model, but we can try to mitigate its detrimental effects. Of primary concern is the fact that the decision boundary will potentially settle into an area which features false positives, as the non-determinism will ensure a mixing of positive and negative values, or even worse, the boundary will try to "memorize" the data, thus producing strange, warped, and/or inaccurate topology.

The classical way of addressing memorization is to train longer on even more data with regularization of some kind. This was going to be done regardless however.

Weighting of data samples seems an attractive and relatively simple approach to the problem. Potentially positive labels can be weighted according to an annealing schedule, whereas negative labels keep a maximum weight of 1 for all of training. This should bias the model away from false positives, likely with a cost of false negatives. However, given the nature of this application, false negatives are equivalent to losing access to parts of the optimization space, which will likely be perfectly fine for current use-case scenarios. A loss of the potentially true optimum in exchange for guaranteed system stability is a necessary price to pay for safety-critical systems.

**Comment by hatfield-c** [comment]
For now, stick to classical non-determinism mitigation methods (i.e. regularization, restricting data domain, etc). Weighting can be attempted if time permits.

**Comment by hatfield-c** [comment]
A means of choosing search directions/locations is necessary. Some information that can be worked with:
- Value of N at a point p. 
    - This gives the pseudo-distance to N's decision boundary from P.
    - It is unknown what all properties this pseudo-distance has which can be exploited. 
    - Ideally the pseudo-distance would be a measure of distance to the nearest decision boundary line/plane in P.
- Gradient of N at a point p.
    - This gives the change in pseudo-distance along the inputs of the neural network.
    - Ideally can be used to determine if the next step will reduce the p-distance by a degree that is considered too much.
- Hessian of N at a point p.
    - This gives the change in the change of the p-distance along the inputs of the neural network
    - Can potentially give information about convexity and the topology of the decision boundary

**Comment by hatfield-c** [comment]
An evaluation function for arbitrary metrics can be integrated with N in some way to allow for a more informed direction of search.

The evaluation function q( ) will need to depend on the inputs of N in some way. Otherwise, there will be no informational link between N and q, and we can set the inputs of N to whatever we wish, as it will have no effect on q( ),

A basic evaluation function for the simple scenario is the feedrate. We want it to be as high as possible, so a direct measure of the feedrate parameter is sufficient. However, there might be other information which may be useful for the search process that is worth considering, such as what the next cutter engagement will be. Furthermore, there  can be derived/calculated measures which may be desirable. 

All of these will require a general means of linking to N for the search process.

**Comment by hatfield-c** [comment]
Search proposal: anchor baking
- Create a finite set of k "anchor" points in the domain of N
- Perform an annealing process such that anchor points repel each other up to a distance, and are also repelled by the decision boundary of N
- Each anchor point is connected to other anchor points that meet some metric
- During search, agent maintains a target anchor to move towards.
    - Anchor points and connections do not encode collision information - that is what the neural network is for.
    - Anchor points serve as guides which stake out regions of valid space

A "growing" annealing might be better suited, where new anchor points are instantiated one at a time at the border of existing anchor space. This could make growing network paths easier, so that all anchor points aren't instantiated and annealed in too-tight clusters.

**Comment by hatfield-c** [comment]
A feature of the search baking process is the necessity for a basic continuous search method that doesn't try to optimize, but simply finds any valid path. Thus this method should be pursued first, and optimization can be added to it.

I've been thinking to try and implement a mixture of A* and RRT in the continuous space bounded by N. A* sets down waypoints every so often, which it uses for backtracking.
- This feels very similar to the anchor baking method, which would honestly be much more preferred.

**Comment by hatfield-c** [comment]
This stack overflow talks about using "visibility maps", which seem like a very interesting concept
- https://gamedev.stackexchange.com/questions/132398/pathfinding-in-non-quantized-space

The difficulty with visibility maps is that N may have an arbitrarily complex shape that requires a large amount of memory to generate the maps. Furthermore, this approach is more geared towards single-source single-destination path finding in a static space. This stands in contrast to the desired application of EPCO for active inference in complicated and dynamic spaces that the agent may not have full control over.

Nonetheless, this approach may contain means for constructing the anchor lattice. At the very least it links to an article which gives some useful information to remember about pathfinding in a continuous space.

**Comment by hatfield-c** [comment]
From the previous comment's link, for future use:

Given an open space with polygonal obstacles, the shortest path between two points is either:
- a direct, uninterrupted line from start to finish
- touching the boundaries of at least one of these obstacles

These facts help clarify that when searching through our constraint space, optimality for short shortest path problems will require getting close to the decision boundary. Other problems may not have this optimality condition.

**Comment by hatfield-c** [comment]
Anchor point generation strategy:
- For 100 desired anchor points, start with 10 seed anchors
- Place them randomly in the space
- Use gradient ascent to move their positions into valid positions of at least some distance from a constraint
- Grow 10 anchor points from each seed anchor
    - Line of Sight (LoS): just shoot straight line rays out from the seed point at random angles. See how far you get.
    - Optional: Repeat growth process for children as well
    - Particle approach has the advantage over LoS that it can likely ensure a more even distribution. 
    - LoS for now

**Comment by hatfield-c** [comment]
In EPCO-#18, I uploaded some heatmaps for some classifiers that were trained with different architecture configurations. These heatmaps show the topology of output from N. While traditionally called the "confidence", I have been calling this output the "pseudo-distance".

The reason for my choice of terminology is that I hoped if we limited the architecture of N in certain ways, then we could induce behavior from N such that the output from N would result in a p-field that somewhat mimicked the that of a Voronoi diagram. The output from the final layer would (hopefully) give us the planar distance (projection? I've been calling it pseudo-distance) from the decision boundary.

This last desired feature has been a point of contention given the results of EPCO-#18. None of the attempted architectures resulted in this behavior. 

In order for psuedo-distance behavior to occur, each point in space must be processed by the lower layers such that they all follow the same distance ordering as that in real space (closer points have lower values, farther points have higher values, etc). The final calculated magnitude of the distance isn't relevant - so long as the ordering is correct. This might be why the uncoupled weight normalization seemed to give the closest topology to the desired behavior, as the magnitude was decoupled but not frozen.

This might be better off induced in the loss function somehow, but it might likely require some notion of distance to the decision boundary be given in the training data. This would be less than ideal, as we aren't supposed to know where the decision boundary is.

**Comment by hatfield-c** [comment]
Label smoothing might help with this behavior:
- https://arxiv.org/pdf/1906.02629.pdf

**Comment by hatfield-c** [comment]
I haven't tried label smoothing yet, but weight decoupling combined with l2 normalization did a great job of smoothing thus far. It isn't a pseudo-distance field like I would have hoped, but it does seem to offer a dramatically smoother space than the other previously explored options.

Weight decoupling (aka weight normalization) here:
- https://arxiv.org/abs/1602.07868

*Edit*

Label smoothing also helped dramatically, and improved covnergence rates when paired with a small l2

**Comment by hatfield-c** [comment]
idea:
- Machine learning algorithm that gives you how far in a specific direction you can step without violating the constraints encoded in N
- Takes as input:
    - Various distance values to points that radiate outward from the current point
    - The direction of movement? (Can this be formatted differently?)
    - Various gradients surrounding the current point

**Comment by hatfield-c** [comment]
We might be able to use Newton's Method of 2nd order optimization to quickly find linear translations through the p-field input space that are valid.

The issue with straight gradient descent in the p-field is that choosing a good step size becomes very difficult when trying to deal with universal conditions. The potential for overstepping, or choosing a linear path that teleports through an invalid region will prove quite problematic, especially for problems that involve continuous dynamic path planning.

A 2nd order method like Newton's Method might in theory allow us to examine potential linear paths by guiding the proposed path endpoint straight to a decision boundary in a direction of interest. The dot product of the gradient at a point with a particular direction gives us the rate of change for that direction. Typically one can just choose the gradient during descent optimization, because it gives the direction of steepest descent. However I am hopeful that either all or most of the rules and conditions chosen for descent optimization will also work with non-steepest direction.

That is, we can choose k uniformly distributed directions from a particular point q, and use Newton's Method to converge the endpoints lines in these directions to the first decision boundary they intersect. 
- If a cilia line detects that it is going uphill, it will revert to hill climbing behavior, and then it will go back to descent behavior once over the hill

**Comment by hatfield-c** [comment]
I'm going to pause smoothing development, as I'm not quite sure how beneficial it will be.

A blind guess-and-check approach will work for a proof of concept for this methodology.

**Comment by hatfield-c** [comment]
This issue will now be closed, as future work on the planner has been moved to EPCO-#25.
